pandas/tests/test_frame.py

```diff
@@ -12328,6 +12328,24 @@ def test_unstack_dtypes(self):
         expected = Series({'float64' : 2, 'object' : 2})
         assert_series_equal(result, expected)
 
+        # GH7405
+        for c, d in (np.zeros(5), np.zeros(5)), \
+                    (np.arange(5, dtype='f8'), np.arange(5, 10, dtype='f8')):
+
+            df = DataFrame({'A': ['a']*5, 'C':c, 'D':d,
+                            'B':pd.date_range('2012-01-01', periods=5)})
+
+            right = df.iloc[:3].copy(deep=True)
+
+            df = df.set_index(['A', 'B'])
+            df['D'] = df['D'].astype('int64')
+
+            left = df.iloc[:3].unstack(0)
+            right = right.set_index(['A', 'B']).unstack(0)
+
+            self.assertEqual(left.shape, (3, 2))
+            tm.assert_frame_equal(left, right, check_dtype=False)
```

Comment:
this is the copy paste of https://github.com/pydata/pandas/issues/7405. 

i am deliberately not casting `right`, because the https://github.com/pydata/pandas/issues/7405 says `unstack` breaks if there is partial selection and mixed data types. so, the `right` frame, (the expected correct result), is the same operations done on a full frame with no mixed data type.